helper/schema/grpc_provider.go

```diff
@@ -929,6 +929,15 @@ func (s *GRPCProviderServer) ReadResource(ctx context.Context, req *tfprotov5.Re
 			return resp, nil
 		}
 
+		// validate every attribute exists and is not null
+		identityAttrs := newIdentityVal.AsValueMap()
+		for k := range identityBlock.Attributes {
+			if v, ok := identityAttrs[k]; !ok || v.IsNull() || !v.IsWhollyKnown() {
+				resp.Diagnostics = convert.AppendProtoDiag(ctx, resp.Diagnostics, fmt.Errorf("identity data missing for attribute '%s'. This is a bug in the provider, which should be reported in the provider's own issue tracker.", k))
+				return resp, nil
+			}
+		}
```

Comment:
Wondering if we should duplicate this validation in the provider? We know core needs/will have this logic at some point and we don't typically do validation like this in the SDKs. (For example, ReadResource it's invalid to return a `NewState` that has unknown values, which core validates, but the SDKs do not)

----------

If we do decide that we want to validate this in the providers, we might consider doing it at the `terraform-plugin-go` level, so it applies to `terraform-plugin-framework` as well

Reply:
Good call!